library/src/cache/patch_manager.rs

```diff
@@ -810,6 +845,95 @@ mod next_boot_patch_tests {
+
+    // The message that was signed.
+    const INFLATED_PATCH_HASH: &str =
+        "6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b";
+
+    // The base64-encoded signature of `MESSAGE` using the private key corresponding to `PUBLIC_KEY`.
+    const SIGNATURE: &str = "ZGccldv01XqHQ76bXuKV/9EQnNK0Q+reQ9bJHVnGfLldF+BLRx0divgPfKP5Df9BJPA3dw1Z1VortfepmMGebP3kS593l5zoktu9MIepxvRAFWNKE5PDTIIvCL/ddTPEHt6NNCeD6HLOMLzbEX3cFZa+lq3UymGi0aqA5DlXirJBGtopojc9nOXZ22n/qHNZIHEkGcqKbSMSK9oC55whKHnlJTbCXdmSyDc65B4PcgseqJom1riVK3XGW1YMrSpuMAU+CDT7HhdESmI1UtH1bYeBITfRhQztdDTfti2vJTf2Y+lYC99CFiISgD7f1m0KUcC+VnEAMZSYtgxSk6AX2A==";
+
+    #[test]
+    fn fails_if_public_key_is_invalid() -> Result<()> {
+        let temp_dir = TempDir::new("patch_manager")?;
+        let mut manager = PatchManager::new(temp_dir.path().to_path_buf(), Some("not a valid key"));
+
+        manager.add_signed_patch_for_test(&temp_dir, 1, INFLATED_PATCH_HASH, Some(SIGNATURE))?;
+
+        assert!(manager.next_boot_patch().is_none());
+
+        Ok(())
+    }
+
+    #[test]
+    fn returns_none_if_patch_is_missing_expected_signature() -> Result<()> {
+        let temp_dir = TempDir::new("patch_manager")?;
+        let mut manager = PatchManager::new(temp_dir.path().to_path_buf(), Some(PUBLIC_KEY));
+
+        manager.add_signed_patch_for_test(&temp_dir, 1, INFLATED_PATCH_HASH, None)?;
+
+        assert!(manager.next_boot_patch().is_none());
+
+        Ok(())
+    }
+
+    #[test]
+    fn returns_none_if_patch_has_invalid_signature() -> Result<()> {
+        let temp_dir = TempDir::new("patch_manager")?;
+        let mut manager = PatchManager::new(temp_dir.path().to_path_buf(), Some(PUBLIC_KEY));
+
+        // Using MESSAGE as a signature because it is valid base64, but not a valid signature.
+        manager.add_signed_patch_for_test(
+            &temp_dir,
+            1,
+            INFLATED_PATCH_HASH,
+            Some(INFLATED_PATCH_HASH),
+        )?;
+
+        assert!(manager.next_boot_patch().is_none());
+
+        Ok(())
+    }
+
+    #[test]
+    fn returns_patch_if_patch_has_valid_signature() -> Result<()> {
+        let temp_dir = TempDir::new("patch_manager")?;
+        let mut manager = PatchManager::new(temp_dir.path().to_path_buf(), Some(PUBLIC_KEY));
+
+        manager.add_signed_patch_for_test(&temp_dir, 1, INFLATED_PATCH_HASH, Some(SIGNATURE))?;
+
+        assert!(manager.next_boot_patch().is_some());
+        let patch = manager.next_boot_patch().unwrap();
+        assert_eq!(patch.number, 1);
+
+        Ok(())
+    }
+
+    #[test]
+    fn returns_patch_with_arbitrary_signature_if_no_public_key() -> Result<()> {
```

Comment:
Tests I can imagine:
1. no public key + signed patch = boots
1. no public key + unsigned patch = boots
1. no public key + invalid signature = ?
1. public key + signed patch = boots
1. public key + unsigned patch = fails
1. public key + invalid signature = fails

Reply:
> no public key + invalid signature = ?

I'm not sure what it would mean for a signature to be invalid with no public key.

Going through the other items on the list:

- [x] no public key + signed patch = boots (`returns_patch_with_arbitrary_signature_if_no_public_key`)
- [x] no public key + unsigned patch = boots (all previously existing tests)
- [x] public key + signed patch = boots (`returns_patch_if_patch_has_valid_signature`)
- [x] public key + unsigned patch = fails (`returns_none_if_patch_is_missing_expected_signature`)
- [x] public key + invalid signature = fails (`returns_none_if_patch_has_invalid_signature`)

Reply:
> I'm not sure what it would mean for a signature to be invalid with no public key.

It could go either way.  Either any signature failure (e.g parsing the signature) causes patch failure, or the signature is completely ignored when there is no public key.  Both would be valid behaviors.